cmd/wait/runner.go

```diff
@@ -0,0 +1,204 @@
+package wait
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/errors/tenant"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/spf13/cobra"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type runner struct {
+	flag   *flag
+	logger micrologger.Logger
+	stdout io.Writer
+	stderr io.Writer
+}
+
+func (r *runner) Run(cmd *cobra.Command, args []string) error {
+	ctx := context.Background()
+
+	err := r.flag.Validate()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.run(ctx, cmd, args)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *runner) run(ctx context.Context, _ *cobra.Command, _ []string) error {
+	kubeconfig, err := ioutil.ReadFile(r.flag.Kubeconfig)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	config, err := clientcmd.NewClientConfigFromBytes(kubeconfig)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	restConfig, err := config.ClientConfig()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	restConfig.Timeout = time.Second * 10
+
+	k8sClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	{
+		r.logger.LogCtx(ctx, "message", "waiting for cluster API to be reachable")
+
+		o := func() error {
+			_, err := k8sClient.CoreV1().Nodes().List(ctx, v1.ListOptions{})
+			if tenant.IsAPINotAvailable(err) {
+				r.logger.LogCtx(ctx, "message", "API not yet available")
+				return microerror.Mask(err)
+			} else if err != nil {
+				r.logger.LogCtx(ctx, "message", "unexpected error contacting API", "error", err)
+				return microerror.Mask(err)
+			}
+			return nil
+		}
+
+		b := backoff.NewMaxRetries(60, 20*time.Second)
+
+		err = backoff.Retry(o, b)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		r.logger.LogCtx(ctx, "message", "API is now reachable")
+	}
+
+	{
+		r.logger.LogCtx(ctx, "message", "waiting for nodes to be ready")
+
+		o := func() error {
+			nodes, err := k8sClient.CoreV1().Nodes().List(ctx, v1.ListOptions{})
+			if err != nil {
+				r.logger.LogCtx(ctx, "message", "error listing nodes", "error", err)
+				return microerror.Mask(err)
+			}
+			nodeCount := len(nodes.Items)
+			readyCount := 0
+			for _, node := range nodes.Items {
+				for _, condition := range node.Status.Conditions {
+					if condition.Type == "Ready" {
+						if condition.Status == "True" {
+							readyCount++
+						}
+						break
+					}
+				}
+			}
+			if nodeCount < 2 {
```

Comment:
This is pretty arbitrary. We don't know exactly how many nodes will be created so 2 seemed like a good lower limit. We could check for one master and one worker or various other combinations.

Reply:
IIRC sonobuoy requires at least 2, so at least this is aligned